teuthology/task/exec.py

```diff
@@ -45,12 +54,30 @@ def task(ctx, config):
         log.info('Running commands on role %s host %s', role, remote.name)
         for c in ls:
             c.replace('$TESTDIR', testdir)
-            remote.run(
-                args=[
-                    'sudo',
-                    'TESTDIR={tdir}'.format(tdir=testdir),
-                    'bash',
-                    '-c',
-                    c],
-                )
-
+            if retry:
+                with safe_while(sleep=sleep_for_retry, tries=retry,
+                                action="exec_with_retry") as proceed:
+                    while proceed():
+                        proc = remote.run(
+                                  args=[
+                                      'sudo',
+                                      'TESTDIR={tdir}'.format(tdir=testdir),
+                                      'bash',
+                                      '-c',
+                                      c],
+                                  timeout=timeout,
+                                  check_status=False,
+                                  wait=True,
+                                 )
+                        if proc.exitstatus == 0:
+                            break
+            else:
+                remote.run(
+                    args=[
+                        'sudo',
+                        'TESTDIR={tdir}'.format(tdir=testdir),
+                        'bash',
+                        '-c',
+                        c],
+                    timeout=timeout
```

Comment:
what happens when timeout is None here? perhaps better for timeout to default to 0

Reply:
I think the paramiko timeout of 0 means infinite wait, I will check that and come back here, Also I think we are defaulting to some high value when its None, I have to check on both.

Reply:
I think you're trying to use 'timeout' incorrectly here and above, because this parameter means 'socket timeout' (or read/write operation timeout on ssh channel), but what you expect to have is a command execution timeout, this functionality is missing for remote.run.